### PR TITLE
router: Add option to bind ports only when ready

### PR DESCRIPTION
**Pull request by marun**
By default the router binds ports 80 and 443 even if no routing configuration is available.  This may be desirable when haproxy is serving traffic directly to clients.  However, the f5 loadbalancer will treat bound ports as indicating that a backend is ready to receive traffic.  This means that router pods that are starting or restarting may be put into rotation before they are ready to serve the current route state and clients may see 503s for perfectly healthy endpoints.  To avoid this possibility, this change adds an option (ROUTER_BIND_PORTS_AFTER_SYNC/--bind-ports-after-sync) to the router that ensures that ports are bound only when a router instance has synced with the api.

Reference: bug 1383663 [link](https://bugzilla.redhat.com/show_bug.cgi?id=1383663)

cc: @openshift/networking

### CODE REVIEW

**Comment by marun** [comment]
[test]

File: images/router/haproxy/conf/haproxy-config.template [hunk 1]
```diff
@@ -85,6 +85,10 @@ listen stats :1936
     mode http
     # Health check monitoring uri.
     monitor-uri /healthz
+{{ if (and .State .ServiceUnits) }}
+    # This uri being configured indicates that routing configuration has been synchronized.
+    monitor-uri /readyz
```
Comment (knobunc) [inline images/router/haproxy/conf/haproxy-config.template]: What's with the z's?  Should we really use that?  I was being somewhat facetious on the call when I said it.

File: pkg/cmd/infra/router/template.go [hunk 2]
```diff
@@ -86,6 +87,7 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
 	flag.StringVar(&o.ReloadScript, "reload", util.Env("RELOAD_SCRIPT", ""), "The path to the reload script to use")
 	flag.DurationVar(&o.ReloadInterval, "interval", reloadInterval(), "Controls how often router reloads are invoked. Mutiple router reload requests are coalesced for the duration of this interval since the last reload time.")
 	flag.BoolVar(&o.ExtendedValidation, "extended-validation", util.Env("EXTENDED_VALIDATION", "true") == "true", "If set, then an additional extended validation step is performed on all routes admitted in by this router. Defaults to true and enables the extended validation checks.")
+	flag.BoolVar(&o.BindOnlyWhenReady, "bind-only-when-ready", util.Env("BIND_ONLY_WHEN_READY", "") == "true", "Only bind ports when route state has been synchronized")
```
Comment (knobunc) [inline pkg/cmd/infra/router/template.go]: The more recent trend has been to put ROUTER_ in the name (see https://github.com/openshift/openshift-docs/pull/3090 for the complete list)

Should we do that here?

File: images/router/haproxy/conf/haproxy-config.template [hunk 3]
```diff
@@ -85,6 +85,10 @@ listen stats :1936
     mode http
     # Health check monitoring uri.
     monitor-uri /healthz
+{{ if (and .State .ServiceUnits) }}
```
Comment (knobunc) [inline images/router/haproxy/conf/haproxy-config.template]: If there are no services in the system (or that match our shard params) will this still bind?

**Comment by knobunc** [comment]
@smarterclayton PTAL

**Comment by knobunc** [comment]
Flake https://github.com/openshift/origin/issues/11775 [test]

**Comment by knobunc** [comment]
flake https://github.com/openshift/origin/issues/10773 [test]

**Comment by marun** [comment]
As per an irc conversation with @knobunc readiness should denote that route state has been synchronized, even if there is no state available.

File: pkg/router/controller/controller.go [hunk 4]
```diff
@@ -57,6 +60,24 @@ func (c *RouterController) Run() {
...
+		syncedAtLeastOnce := c.RoutesListSuccessfulAtLeastOnce() && c.EndpointsListSuccessfulAtLeastOnce() &&
+			(c.Namespaces == nil || c.filteredByNamespace)
+		if syncedAtLeastOnce {
+			if err := c.Plugin.SetSyncedAtLeastOnce(); err != nil {
+				utilruntime.HandleError(err)
+			} else {
+				return
+			}
+		}
+		time.Sleep(1)
```
Comment (smarterclayton) [inline pkg/router/controller/controller.go]: This is a hot loop, don't do this.  Sleep longer.

File: pkg/router/controller/controller.go [hunk 5]
```diff
@@ -57,6 +60,24 @@ func (c *RouterController) Run() {
...
+
+// watchForFirstSync signals the router when it sees that the various
+// watchers have successfully listed data from the api.
+func (c *RouterController) watchForFirstSync() {
+	for {
+		syncedAtLeastOnce := c.RoutesListSuccessfulAtLeastOnce() && c.EndpointsListSuccessfulAtLeastOnce() &&
+			(c.Namespaces == nil || c.filteredByNamespace)
+		if syncedAtLeastOnce {
+			if err := c.Plugin.SetSyncedAtLeastOnce(); err != nil {
+				utilruntime.HandleError(err)
```
Comment (smarterclayton) [inline pkg/router/controller/controller.go]: invert this condition and return early.

File: pkg/router/controller/controller.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/router/controller/controller.go]: Minimum of 50ms

Comment (marun) [inline pkg/router/controller/controller.go]: Good call, I keep forgetting that ``time.Sleep()`` takes a duration not an int.  My intention was to sleep for 1s.

**Comment by marun** [comment]
I'm still working on this, coordinating readiness and reload is more complicated than I originally thought.

**Comment by marun** [comment]
A router now becomes ready even when no items are queued for either of the watchers.  Some of the integration tests are broken, though.  Will investigate tomorrow.

**Comment by marun** [comment]
It's not possible to configure multiple ``monitor-uri`` paths on a given stats port, so adding a readiness check would require opening multiple stats ports.  I'm assuming that is out of scope here, so this PR no longer includes the addition of a readiness check.

**Comment by knobunc** [comment]
@ramr  @openshift/networking PTAL

**Comment by marun** [comment]
flake #11685 [test]

**Comment by marun** [comment]
flake #11887 re-[test]

File: pkg/router/controller/controller.go [hunk 6]
```diff
@@ -57,6 +62,51 @@ func (c *RouterController) Run() {
...
+
+// handleFirstSync signals the router when it sees that the various
+// watchers have successfully listed data from the api.
+func (c *RouterController) handleFirstSync() bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	synced := c.RoutesListSuccessfulAtLeastOnce() &&
+		c.EndpointsListSuccessfulAtLeastOnce() &&
+		(c.Namespaces == nil || c.filteredByNamespace)
```
Comment (ramr) [inline pkg/router/controller/controller.go]: Is the `c.Namespaces == nil` check needed? Wouldn't `c.filteredByNamespace` suffice? I see `c.filteredByNamespace` set in handle namespaces on no errors.

**Comment by ramr** [comment]
Changes LGTM - just a small question.

File: pkg/router/controller/controller.go (hunk 6 above)
Comment (marun) [inline pkg/router/controller/controller.go]: Yes, it is needed.  If ``c.Namespaces == nil `` then ``c.filteredByNamespace`` will never be set to ``true``.

Comment (marun) [inline pkg/router/controller/controller.go]: If ``c.Namespaces == nil``, then ``c.HandleNamespaces()`` will never be called:

https://github.com/openshift/origin/blob/master/pkg/router/controller/controller.go#L51

**Comment by knobunc** [comment]
[merge]

**Pull request merged**